crates/circuit/src/dag_circuit.rs

```diff
@@ -5945,17 +5892,19 @@ impl DAGCircuit {
     }
 
     /// Returns an iterator over the first layer of the `DAGCircuit``.
-    pub fn front_layer<'a>(&'a self, py: Python) -> Box<dyn Iterator<Item = NodeIndex> + 'a> {
-        let mut graph_layers = self.multigraph_layers(py);
+    pub fn front_layer(&self) -> impl Iterator<Item = NodeIndex> {
+        let mut graph_layers = self.multigraph_layers();
         graph_layers.next();
 
         let next_layer = graph_layers.next();
         match next_layer {
             Some(layer) => Box::new(layer.into_iter().filter(|node| {
                 matches!(self.dag.node_weight(*node).unwrap(), NodeType::Operation(_))
-            })),
-            None => Box::new(vec![].into_iter()),
+            }))
+            .collect(),
+            None => vec![],
```

Comment:
Is there any reason we're ditching the iterators here?

Reply:
I mostly just don't like using `Box` unless it's an absolute must, and while updating this signature, I noticed that this function just emits from a single `Vec` (the second one) of `multigraph_layers`, so doing this doesn't interrupt any sort of lazy iteration that we might otherwise benefit from.

But, we can actually do this without the `collect` and without disrupting any would-be iterators, which I've just done now in 1254ba1 😄.